File: app/src/main/java/com/privat/contacts/presentation/users/UsersFragment.java
package com.privat.contacts.presentation.users;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.privat.contacts.R;
import com.privat.contacts.databinding.FragmentUsersBinding;
import com.privat.contacts.presentation.base.BaseMvpView;
import com.privat.contacts.presentation.host.BottomNavigationHostFragmentDirections;
import com.privat.contacts.presentation.users.adapter.UsersAdapter;
import com.privat.contacts.presentation.users.model.UserItemUi;

import java.util.List;

import javax.inject.Inject;

public class UsersFragment extends BaseMvpView<UsersPresenter> implements UsersView {
    @Inject
    public UsersPresenter usersPresenter;
    private FragmentUsersBinding binding;
    private final UsersAdapter usersAdapter = new UsersAdapter(onClickId -> {
        usersPresenter.openUser(onClickId);
    });
    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container, Bundle savedInstanceState) {
        binding = FragmentUsersBinding.inflate(inflater, container, false);
        View root = binding.getRoot();
        ((AppCompatActivity) requireActivity()).getSupportActionBar().setTitle(R.string.all_users);
        DividerItemDecoration itemDecorator = new DividerItemDecoration(requireContext(), DividerItemDecoration.VERTICAL);
        itemDecorator.setDrawable(ContextCompat.getDrawable(requireContext(), R.drawable.divider));
        binding.rvUsers.setLayoutManager(new LinearLayoutManager(requireContext()));
        binding.rvUsers.setAdapter(usersAdapter);
        binding.rvUsers.addItemDecoration(itemDecorator);
        return root;
    }
    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        binding.fabFetchNewUser.setOnClickListener(listener -> {
            usersPresenter.fetchNewUser();
        });
        usersPresenter.clearTempData();
        usersPresenter.startObserve();
    }

    @Override
    public void onStart() {
        super.onStart();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        usersPresenter.stopObserve();
        binding = null;
    }
    @Override
    public void showUsers(List<UserItemUi> users) {
        if (binding != null) {
            usersAdapter.updateData(users);
            binding.rvUsers.smoothScrollToPosition(users.size() - 1);
        }
    }
    @Override
    public void navigateToUser(int userId) {
        Navigation.findNavController(requireActivity(), R.id.nav_host_fragment_activity_main)
                .navigate(
                        BottomNavigationHostFragmentDirections
                                .actionBottomNavigationHostFragmentToDetailsFragment(userId));
    }
    @Override
    protected UsersPresenter presenter() {
        return usersPresenter;
    }
}

File: app/src/main/java/com/privat/contacts/presentation/users/UsersFragment.kt
package com.privat.contacts.presentation.users

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.navigation.Navigation.findNavController
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import com.privat.contacts.R
import com.privat.contacts.databinding.FragmentUsersBinding
import com.privat.contacts.presentation.host.BottomNavigationHostFragmentDirections
import com.privat.contacts.presentation.users.model.UserItemUi
import dagger.android.support.DaggerFragment
import javax.inject.Inject

class UsersFragment : DaggerFragment(), UsersView {
    @Inject
    lateinit var usersPresenter: UsersPresenter
    private var _binding: FragmentUsersBinding? = null
    private val binding get() = _binding!!
    private val usersAdapter = UsersAdapter { onClickId: Int -> usersPresenter.openUser(onClickId) }
    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?, savedInstanceState: Bundle?): View? {
        _binding = FragmentUsersBinding.inflate(inflater, container, false)
        val root: View = binding.root
        binding.rvUsers.layoutManager = LinearLayoutManager(requireContext())
        binding.rvUsers.adapter = usersAdapter
        (requireActivity() as AppCompatActivity).supportActionBar!!.setTitle(R.string.all_users)
        val itemDecorator = DividerItemDecoration(requireContext(), DividerItemDecoration.VERTICAL)
        itemDecorator.setDrawable(ContextCompat.getDrawable(requireContext(), R.drawable.divider)!!)
        binding.rvUsers.addItemDecoration(itemDecorator)
        return root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.fabFetchNewUser.setOnClickListener { listener: View? -> usersPresenter.fetchNewUser() }
        usersPresenter.clearTempData()
    }

    override fun onStart() {
        super.onStart()
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        usersPresenter.onAttach(this)
    }

    override fun onDetach() {
        super.onDetach()
        usersPresenter.onDetach()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun showUsers(users: List<UserItemUi>) {
        usersAdapter.updateData(users)
        usersAdapter.notifyDataSetChanged()
    }

    override fun navigateToUser(userId: Int) {
        findNavController(requireActivity(), R.id.nav_host_fragment_activity_main)
                .navigate(
                        BottomNavigationHostFragmentDirections
                                .actionBottomNavigationHostFragmentToDetailsFragment(userId))
    }
}